src/logic/equiv/fin.lean

```diff
@@ -424,6 +425,56 @@ def fin_prod_fin_equiv : fin m × fin n ≃ fin (m * n) :=
         ... = y.1 : nat.mod_eq_of_lt y.2),
   right_inv := λ x, fin.eq_of_veq $ nat.mod_add_div _ _ }
 
+/-- The equivalence between `a` and `(a / n, a % n)` for nonzero `n`.
+
+This is like `fin_prod_fin_equiv.symm` but with `m` infinite. -/
+@[simps]
+def nat.div_mod_equiv (n : ℕ) [ne_zero n] : ℕ ≃ ℕ × fin n :=
+{ to_fun := λ a, (a / n, fin.of_nat' a),
+  inv_fun := λ p, p.1 * n + p.2,  -- TODO: is there a canonical order of `*` and `+` here?
+  left_inv := λ a, nat.div_add_mod' _ _,
+  right_inv := λ p, begin
+    refine prod.ext _ (fin.ext $ nat.mul_add_mod_of_lt p.2.is_lt),
+    dsimp only,
+    rw [add_comm, nat.add_mul_div_right _ _ (ne_zero.pos n), nat.div_eq_of_lt p.2.is_lt, zero_add],
+  end }
+
+/-- The equivalence between `a` and `(a / n, a % n)` for nonzero `n`. -/
+@[simps]
+def int.div_mod_equiv (n : ℕ) [ne_zero n] : ℤ ≃ ℤ × fin n :=
+{ to_fun := λ a, (a / n, (a.nat_mod n)),
+  inv_fun := λ p, p.1 * n + (p.2 : ℕ),
+  left_inv := λ a, begin
+    dsimp only,
+    convert int.div_add_mod' _ (n : ℤ),
+    rw fin.coe_coe_of_lt ,
+    { rw [int.nat_mod, int.to_nat_of_nonneg],
+      refine int.mod_nonneg _ (nat.cast_ne_zero.mpr $ ne_zero.ne n), },
+    rw ←int.coe_nat_lt,
+    { rw [int.nat_mod,
+        int.to_nat_of_nonneg (int.mod_nonneg _ (nat.cast_ne_zero.mpr $ ne_zero.ne n))],
+      refine int.mod_lt_of_pos _ _,
+      rw [←int.coe_nat_zero, int.coe_nat_lt],
+      exact ne_zero.pos n, },
+  end,
+  right_inv := λ ⟨q, ⟨r, hrn⟩⟩, begin
+    -- TODO: either we're missing some important lemmas here, or they're not available in this file.
+    dsimp [←fin.of_nat'_eq_coe, fin.of_nat'],
+    have hn : (n : ℤ) ≠ 0 := (nat.cast_ne_zero.mpr $ ne_zero.ne n),
+    congr,
+    { rw [int.add_div_of_dvd_left (dvd_mul_left _ _), int.mul_div_cancel _ hn,
+        int.div_eq_zero_of_lt (int.coe_nat_nonneg r) (int.coe_nat_lt.mpr hrn), add_zero] },
+    { apply int.coe_nat_inj,
+      rw [int.nat_mod, nat.mod_eq_of_lt,
+        int.to_nat_of_nonneg (int.mod_nonneg _ (nat.cast_ne_zero.mpr $ ne_zero.ne n)), add_comm,
+        int.add_mul_mod_self, int.mod_eq_of_lt  (int.coe_nat_nonneg r) (int.coe_nat_lt.mpr hrn)],
+      rw ←int.coe_nat_lt,
+      rw [int.to_nat_of_nonneg (int.mod_nonneg _ (nat.cast_ne_zero.mpr $ ne_zero.ne n))],
+      refine int.mod_lt_of_pos _ _,
+      rw [←int.coe_nat_zero, int.coe_nat_lt],
+      exact ne_zero.pos n, }
```

Comment:
Only slightly shorter. Had to inline `key₂` and we're missing other `norm_cast` lemmas because of imports which would have simplified this a bit further.
```suggestion
    dsimp [←fin.of_nat'_eq_coe, fin.of_nat'],
    rw [add_comm _ (r : ℤ), mul_comm q _],
    have key₁ := (@int.div_mod_unique _ n r q (int.coe_nat_lt_coe_nat_of_lt (ne_zero.pos n))).mpr
      ⟨rfl, int.coe_nat_le_coe_nat_of_le (zero_le r), int.coe_nat_lt_coe_nat_of_lt hrn⟩,
    refine prod.ext key₁.1 (fin.ext _),
    simp only [fin.coe_mk],
    apply @nat.cast_injective ℤ,
    have key₂ : ∀ a : ℤ, (a.nat_mod n : ℤ) = a % n := λ a, int.to_nat_of_nonneg
      (a.mod_nonneg (by exact_mod_cast (ne_zero.out : n ≠ 0) : (n : ℤ) ≠ 0)),
    simp only [key₂, int.coe_nat_mod, int.add_mul_mod_self_left, int.mod_mod],
    exact int.mod_eq_of_lt (int.coe_nat_le_coe_nat_of_le (zero_le r))
      (int.coe_nat_lt_coe_nat_of_lt hrn),
```
optionally, you could pull these out into a separate file so that we can simplify and not screw up imports. That would also prevent needing to forward-port this file.

Reply:
Thanks for the golfs! Do you have a suggestion for which file would be more appropriate to put this in, or what the new name should be?

Reply:
I took the other golg instead, sorry!